# A Brief COBRA TOOLBOX 101

<img src="figures/cobralogo.jpg" width="700">

## From microbe to model
A constraint-based metabolic model contains the stoichiometric matrix (S) with reactions and metabolites. S is a stoichiometric representation of metabolic networks corresponding to the reactions in the biochemical pathway. In each column of the S matrix is a biochemical reaction (n) and in each row is a precise metabolite (m). There is a stoichiometric coefficient of zero, which means that metabolite does not participate in that distinct reaction. The coefficient also can be positive when the appropriate metabolite is produced, or negative for every metabolite consumed.

<img src="figures/S-matrix.png" width="500">


## Begin by initializing the Cobra toolbox

Initialize the Cobra Toolbox using the |initCobraToolbox| function.

In [12]:
initCobraToolbox



      _____   _____   _____   _____     _____     |
     /  ___| /  _  \ |  _  \ |  _  \   / ___ \    |   COnstraint-Based Reconstruction and Analysis
     | |     | | | | | |_| | | |_| |  | |___| |   |   The COBRA Toolbox - 2017
     | |     | | | | |  _  { |  _  /  |  ___  |   |
     | |___  | |_| | | |_| | | | \ \  | |   | |   |   Documentation:
     \_____| \_____/ |_____/ |_|  \_\ |_|   |_|   |   http://opencobra.github.io/cobratoolbox
                                                  | 

 > Checking if git is installed ...  Done.
 > Checking if the repository is tracked using git ...  Done.
 > Checking if curl is installed ...  Done.
 > Checking if remote can be reached ...  Done.
 > Initializing and updating submodules ... Done.
 > Define CB map output... set to svg.
 > Retrieving models ...   Done.
 > TranslateSBML is installed and working properly.
 > Configuring solver environment variables ...
   - ILOG_CPLEX_PATH:  --> set this path manually after installing the solver
  

### Next, load a load the model into the directory

To load a model from a MAT-file, you can simply use load('the filename (with or without file extension))'. 

In [13]:
% load model
load('Model_iJB785.mat');

Once the model is loaded it can be used directly with The COBRA Toolbox functions. To view the data stored in the model use the following command.

If you have matlab, use the command |open model|, otherwise open the model by typing the variable name

In [14]:
model


model = 

           mets: {769x1 cell}
       metNames: {769x1 cell}
    metFormulas: {768x1 cell}
          genes: {785x1 cell}
        grRules: {851x1 cell}
           rxns: {851x1 cell}
       rxnNames: {851x1 cell}
     subSystems: {851x1 cell}
         csense: 'EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE...'
              S: [769x851 double]
             lb: [851x1 double]
             ub: [851x1 double]
              b: [769x1 double]
              c: [851x1 double]
            rev: [851x1 int32]
    description: 'iJB785'
          rules: {851x1 cell}
     rxnGeneMat: [851x785 double]



You can see that there are 769 metabolites, 785 genes, and 851 reactions in the model names iJB785.
It is required for a model to consist of the descriptive fields: model.mets and model.rxns, which represent the metabolites and the reactions respectively.

In [28]:
model.rxns


ans = 

    'QULNS'
    'ORNDC'
    'LEUTA'
    'MSBENZMT'
    'DESAT18a'
    'FUM'
    'PHYFXOR'
    'VPAMTr'
    'GLYCL'
    'NDPK7'
    'GTPCI'
    'MTHFC'
    'LYCBC2'
    'ORNTA'
    'UPPDC1'
    'GARFT'
    'H4THDPR'
    'UDPG4E'
    'GLCS3'
    'ASPTA'
    'PRAGSr'
    'ACKr'
    'KAS14'
    'UDCPDPS'
    'GLNS'
    'SHKK'
    'G1PACT'
    '5DOAN'
    'OHPBAT'
    'Htex'
    '3HAD60'
    'PGI'
    'PNTK'
    '3OAR60'
    'ACGK'
    'LEUabcpp'
    'G5SD'
    'DTMPK'
    'NOR'
    'NDH-1(1)um'
    'AOOA'
    'CMPA'
    'U23GAAT2'
    'PGK'
    'IMPD'
    'LPOR'
    'CHRPL'
    'GCALDDy'
    'RNDR1'
    'AIRC2'
    'MTAP'
    'HEMEAS'
    'DHORD3um'
    'TRPS1'
    'NNATr'
    '3OAR120'
    'HISTDb'
    'SPS'
    'ADCL'
    'CYTK1'
    'PSIum'
    'Cobalt2abcppI'
    'THRS'
    'AOXPBDC'
    'NAMNPP'
    'PERD'
    'DAPDC'
    'COCHL'
    '3OAR100'
    'PFK_3'
    'PPK2r'
    'ZNabcpp'
    'GLYDHDA'
    'ANS'
    'SUCBZL'
    'ADPT'
    'RNDR3'
    'DHNANT'
    'GND'
    'O2tcx'
 

    'HISTRS'
    'ASNTRS'
    'ASPTRS'
    'THRTRS'
    'LEUTRS'
    'ILETRS'
    'LYSTRS'
    'ALATRS'
    'VALTRS'
    'FMETTRS'
    'EX_photon410_e'
    'EX_photon430_e'
    'EX_photon450_e'
    'EX_photon470_e'
    'EX_photon490_e'
    'EX_photon510_e'
    'EX_photon530_e'
    'EX_photon550_e'
    'EX_photon570_e'
    'EX_photon590_e'
    'EX_photon610_e'
    'EX_photon630_e'
    'EX_photon650_e'
    'EX_photon670_e'
    'EX_photon690_e'
    'ZXANHX'
    'CXANHX'
    'PHOA410um'
    'PHOA430um'
    'PHOA450um'
    'PHOA470um'
    'PHOA490um'
    'PHOA510um'
    'PHOA530um'
    'PHOA550um'
    'PHOA570um'
    'PHOA590um'
    'PHOA610um'
    'PHOA630um'
    'PHOA650um'
    'PHOA670um'
    'PHOA690um'
    'DM_pho_loss_c'
    'ETCHLAS'
    'ETBCARO'
    'ETCPC'
    'STACPSI'
    'A5PISO'
    'KDOPS'
    'KDOPP'
    'KDOCT2'
    'MOAT'
    'UAG2E'
    'ACGAMT'
    'ACMAMT'
    'ICLIPAS'
    'ICLIPAabcpp'
    'OANTS'
    'OANTIabcpp'
    'OANTILpp'
    'COLIPAabcex'
    'MPTG'
    'MPTG2

The fields in a COBRA model are commonly column vectors, which is important to note when writing functions manipulating these fields.
There are a few more fields present in each COBRA model:
model.lb, indicating the lower bounds of each reaction, and model.ub indicating the upper bound of a reaction. 

In [29]:
% this displays an array with reaction names and flux bounds.
[{'Reaction ID', 'Lower Bound', 'Upper Bound'};...   
 model.rxns, num2cell(model.lb), num2cell(model.ub)]


ans = 

    'Reaction ID'          'Lower Bound'    'Upper Bound'
    'QULNS'                [          0]    [ 1.0000e+09]
    'ORNDC'                [          0]    [ 1.0000e+09]
    'LEUTA'                [-1.0000e+09]    [ 1.0000e+09]
    'MSBENZMT'             [          0]    [ 1.0000e+09]
    'DESAT18a'             [-1.0000e+09]    [ 1.0000e+09]
    'FUM'                  [-1.0000e+09]    [ 1.0000e+09]
    'PHYFXOR'              [          0]    [ 1.0000e+09]
    'VPAMTr'               [-1.0000e+09]    [          0]
    'GLYCL'                [          0]    [ 1.0000e+09]
    'NDPK7'                [-1.0000e+09]    [ 1.0000e+09]
    'GTPCI'                [          0]    [ 1.0000e+09]
    'MTHFC'                [-1.0000e+09]    [ 1.0000e+09]
    'LYCBC2'               [-1.0000e+09]    [ 1.0000e+09]
    'ORNTA'                [          0]    [          0]
    'UPPDC1'               [          0]    [ 1.0000e+09]
    'GARFT'                [          0]    [ 1.0000e+09]
    '

    'DHAD1'                [          0]    [ 1.0000e+09]
    'IPPS'                 [          0]    [ 1.0000e+09]
    'PPBNGS'               [          0]    [ 1.0000e+09]
    'PRE3BS'               [          0]    [ 1.0000e+09]
    'PHYTES2'              [          0]    [ 1.0000e+09]
    '3HAD80'               [-1.0000e+09]    [ 1.0000e+09]
    'LTHRK'                [          0]    [ 1.0000e+09]
    'PPS'                  [          0]    [ 1.0000e+09]
    'ADCPS2'               [          0]    [ 1.0000e+09]
    'GF6PTA'               [-1.0000e+09]    [ 1.0000e+09]
    'O2tpp'                [-1.0000e+09]    [ 1.0000e+09]
    'PGMT'                 [-1.0000e+09]    [ 1.0000e+09]
    '3OAR140'              [-1.0000e+09]    [ 1.0000e+09]
    'NMNS'                 [-1.0000e+09]    [ 1.0000e+09]
    'NDPK2'                [-1.0000e+09]    [ 1.0000e+09]
    'AGDI'                 [          0]    [ 1.0000e+09]
    'LPADSS2'              [          0]    [ 1.0000e+09]
    'DPR'     

    'KARI_23dhmp'          [          0]    [ 1.0000e+09]
    'IPMD'                 [          0]    [ 1.0000e+09]
    '3PGtcx'               [          0]    [ 1.0000e+09]
    'CDPMEK'               [          0]    [ 1.0000e+09]
    'CITCIa'               [-1.0000e+09]    [ 1.0000e+09]
    'SUCDum'               [-1.0000e+09]    [ 1.0000e+09]
    'DVOCHR'               [          0]    [ 1.0000e+09]
    'H2CO3_NAt_syn'        [-1.0000e+09]    [ 1.0000e+09]
    'HEMEOS'               [-1.0000e+09]    [ 1.0000e+09]
    'ALAD_L'               [-1.0000e+09]    [ 1.0000e+09]
    'RPE'                  [-1.0000e+09]    [ 1.0000e+09]
    'MPOMT'                [          0]    [ 1.0000e+09]
    'NDHPQRum'             [          0]    [ 1.0000e+09]
    'UAMAS'                [          0]    [ 1.0000e+09]
    'CITCIb'               [-1.0000e+09]    [ 1.0000e+09]
    'BPNT2'                [          0]    [ 1.0000e+09]
    'BM_COFACTORS'         [          0]    [ 1.0000e+09]
    'DCTPD2'  

    'G1SAT'                [-1.0000e+09]    [ 1.0000e+09]
    'ILETA'                [-1.0000e+09]    [ 1.0000e+09]
    'PDS1'                 [          0]    [ 1.0000e+09]
    'UAAGDS'               [          0]    [ 1.0000e+09]
    'GLCDBRAN3'            [          0]    [ 1.0000e+09]
    'HEX1'                 [          0]    [ 1.0000e+09]
    'ATPSum'               [-1.0000e+09]    [ 1.0000e+09]
    'CHPHYS'               [          0]    [ 1.0000e+09]
    'BCAROHX'              [          0]    [ 1.0000e+09]
    'GLYCLTDx'             [          0]    [ 1.0000e+09]
    'PIuabcpp'             [          0]    [ 1.0000e+09]
    'PC20M'                [          0]    [ 1.0000e+09]
    'ADSS'                 [          0]    [ 1.0000e+09]
    'BM_RNA'               [          0]    [ 1.0000e+09]
    'DM_dialurate_c'       [          0]    [ 1.0000e+09]
    'THRPD'                [          0]    [ 1.0000e+09]
    'TMDS3'                [          0]    [ 1.0000e+09]
    'DHFR'    

    'ENOPH'                [          0]    [ 1.0000e+09]
    'ARD'                  [          0]    [ 1.0000e+09]
    'UNK3'                 [          0]    [ 1.0000e+09]
    'PHYPQOX'              [          0]    [ 1.0000e+09]
    'ZISO'                 [          0]    [ 1.0000e+09]
    'ZCARDS'               [          0]    [ 1.0000e+09]
    'PLYCOI'               [          0]    [ 1.0000e+09]
    'CHLADMT'              [          0]    [ 1.0000e+09]
    'PHEOASE'              [          0]    [ 1.0000e+09]
    'PHEOX'                [          0]    [ 1.0000e+09]
    'PHYTK'                [          0]    [ 1.0000e+09]
    'TRNFE'                [-1.0000e+09]    [ 1.0000e+09]
    'PTHPS'                [          0]    [ 1.0000e+09]
    'SPR'                  [          0]    [ 1.0000e+09]
    'THBTGT'               [          0]    [ 1.0000e+09]
    'NADDP'                [          0]    [ 1.0000e+09]
    'CYSDS'                [          0]    [ 1.0000e+09]
    'THISAT'  

    'ZXANHX'               [          0]    [ 1.0000e+09]
    'CXANHX'               [          0]    [ 1.0000e+09]
    'PHOA410um'            [          0]    [ 1.0000e+09]
    'PHOA430um'            [          0]    [ 1.0000e+09]
    'PHOA450um'            [          0]    [ 1.0000e+09]
    'PHOA470um'            [          0]    [ 1.0000e+09]
    'PHOA490um'            [          0]    [ 1.0000e+09]
    'PHOA510um'            [          0]    [ 1.0000e+09]
    'PHOA530um'            [          0]    [ 1.0000e+09]
    'PHOA550um'            [          0]    [ 1.0000e+09]
    'PHOA570um'            [          0]    [ 1.0000e+09]
    'PHOA590um'            [          0]    [ 1.0000e+09]
    'PHOA610um'            [          0]    [ 1.0000e+09]
    'PHOA630um'            [          0]    [ 1.0000e+09]
    'PHOA650um'            [          0]    [ 1.0000e+09]
    'PHOA670um'            [          0]    [ 1.0000e+09]
    'PHOA690um'            [          0]    [ 1.0000e+09]
    'DM_pho_lo

### Next, 

<!--NAVIGATION-->
< [Original Simulation published in Broddrick et al 2016 PNAS](Original-Simulation-Broddrick-2016-PNAS.ipynb)>